crates/jxl-frame/src/data/spline.rs

```diff
@@ -1,26 +1,203 @@
+            }
+        }
+        for i in &mut self.sigma_dct {
+            *i = unpack_signed(decoder.read_varint(bitstream, 5)?);
+        }
+        Ok(())
+    }
+
+    // TODO check Maximum total_estimated_area_reached
+    fn dequant(
+        &self,
+        start_point: (i32, i32),
+        quant_adjust: i32,
+        base_correlation_x: f32,
+        base_correlation_b: f32,
+        // estimated_area_reached: &mut usize,
+    ) -> Spline {
+        // let mut manhattan_distance = 0;
+        let mut points = Vec::with_capacity(self.points_deltas.len() + 1);
+
+        let mut cur_value = start_point;
+        points.push((cur_value.0 as f32, cur_value.1 as f32));
+        let mut cur_delta = (0, 0);
+        for delta in &self.points_deltas {
+            cur_delta.0 += delta.0;
+            cur_delta.1 += delta.1;
+            // manhattan_distance += cur_delta.0.abs() + cur_delta.1.abs();
+            cur_value.0 += cur_delta.0;
+            cur_value.1 += cur_delta.1;
+            points.push((cur_value.0 as f32, cur_value.1 as f32));
+        }
+
+        let mut xyb_dct = [[0f32; 32]; 3];
+        let mut sigma_dct = [0f32; 32];
+        let mut width_estimate = 0f32;
+
+        let quant_adjust = quant_adjust as f32;
+        let inverted_qa = if quant_adjust >= 0.0 {
+            1.0 / (1.0 + quant_adjust / 8.0)
+        } else {
+            1.0 - quant_adjust / 8.0
+        };
+
+        const CHANNEL_WEIGHTS: [f32; 4] = [0.0042, 0.075, 0.07, 0.3333];
+        for chan_idx in 0..2 {
+            for i in 0..32 {
+                xyb_dct[chan_idx][i] =
+                    self.xyb_dct[chan_idx][i] as f32 * CHANNEL_WEIGHTS[chan_idx] * inverted_qa;
+            }
+        }
+        for i in 0..32 {
+            xyb_dct[0][i] += base_correlation_x * xyb_dct[1][i];
+            xyb_dct[2][i] += base_correlation_b * xyb_dct[1][i];
+        }
+        for i in 0..32 {
+            sigma_dct[i] = self.sigma_dct[i] as f32 * CHANNEL_WEIGHTS[3] * inverted_qa;
+            // let weight = (self.sigma_dct[i]).abs() as f32 * (inverted_qa).ceil();
+            // width_estimate += weight * weight;
+        }
+
+        // *estimated_area_reached += (width_estimate * manhattan_distance as f32) as usize;
+
+        Spline {
+            points,
+            xyb_dct,
+            sigma_dct,
+        }
+    }
+}
 
-        let num_splines = decoder.read_varint(bitstream, 2)? + 1;
-        let mut last = (0i32, 0i32);
-        todo!()
+impl ToString for Spline {
+    fn to_string(&self) -> String {
+        let mut res = String::from("Spline\n");
+        for i in self.xyb_dct.iter().chain(&[self.sigma_dct]) {
+            for val in i {
+                write!(res, "{} ", val).unwrap();
+            }
+            writeln!(res).unwrap();
+        }
+        for point in &self.points {
+            writeln!(res, "{} {}", point.0 as i32, point.1 as i32).unwrap();
+        }
+        writeln!(res, "EndSpline").unwrap();
+        res
     }
 }
```

Comment:
`impl Display` would be the better, as it allows using values in `format!` and we get `impl ToString` for free.

Reply:
Thanks!